book/thinking_in_nu.md

```diff
@@ -1,95 +1,432 @@
+
+To make sure the first line is _displayed_, use the [`print` command](/commands/docs/print.md):
+
+```nu
+def latest-file [] {
+print "Returning last file"
+ls | sort-by modified | last
+}
+```
+
+Also compare:
+
+```nu
+40; 50; 60
+```
+
+::: tip
+A semicolon is the same as a newline in a Nushell expression. The above is the same as a file or multi-line command:
+
+```nu
+40
+50
+60
+```
+
+or
+
+```nu
+echo 40
+echo 50
+echo 60
+```
+
+:::
+
+In all of the above:
+
+- The first value is evaluated as the integer 40 but is not returned
+- The second value is evaluated as the integer 50 but is not returned
+- The third value is evaluated as the integer 60, and since it is the last
+  value, it is is returned and displayed (rendered).
+
+::: warning Thinking in Nushell
+When debugging unexpected results, be on the lookout for:
+
+- Subexpressions (e.g., commands or pipelines) that ...
+- ... output a (non-`null`) value ...
+- ... where that value isn't returned from the parent expression.
+
+These can be likely sources of issues in your code.
+:::
+
+## Every Command Returns a Value
+
+Some languages have the concept of "statements" which don't return values. Nushell does not.
+
+In Nushell, **_every command returns a value_**, even if that value is `null` (the `nothing` type). Consider the following multiline expression:
+
+```nu:line-numbers
+let p = 7
+print $p
+$p * 6
+```
+
+1. Line 1: The integer 5 is assigned to `$p`, but the return value of the
+   [`let` command](/commands/docs/let.md) itself is `null`. However, because it is not the last
+   value in the expression, it is not displayed.
+2. Line 2: The return value of the `print` command itself is `null`, but the `print` command
+   forces its argument (`$p`, which is 5) to be _displayed_. As with Line 1, the `null` return value
+   is discarded since this isn't the last value in the expression.
+3. Line 3: Evaluates to the integer value 42. As the last value in the expression, this is the return
+   result, and is also displayed (rendered).
+
+::: warning Thinking in Nushell
+Becoming familiar with the output types of common commands will help you understand how
+to combine simple commands together to achieve complex results.
+
+`help <command>` will show the signature, including the output type(s), for each command in Nushell.
+:::
 
 ## Think of Nushell as a Compiled Language
 
-An important part of Nushell's design and specifically where it differs from many dynamic languages is that Nushell converts the source you give it into something to run, and then runs the result. It doesn't have an `eval` feature which allows you to continue pulling in new source during runtime. This means that tasks like including files to be part of your project need to be known paths, much like includes in compiled languages like C++ or Rust.
+In Nushell, there are exactly two, high-level stages when running code:
```

Comment:
I agree - The *detail* behind this is covered in the *"How Nushell Code Gets Run"* Chapter, but I wanted to be more clear about the benefits and limitations here.  I moved the `eval` limitation above the "Features" to give it a bit more visibility.  Also revised it to bring in more of the 'static' terminology here.

I think this chapter has always been problematic because we really don't want people think about Nushell as a *compiled* language as much as a *static* one.  However, we've been using that phrase for so long, I think we're stuck with it.

One problem is that the "How Nushell Code Gets Run" chapter talks about how Nushell's Parsing into IR is similar to Python's parsing to bytecode.  If they are both *compiling*, then it's tough to make a case for calling Nushell a "Compiled Language" but not Python.

I've tried to rewrite as much of it as possible to place the focus on *"Static vs. Dynamic"* instead.

Reply:
Yeah, the word "compiling" is very general and can fit to almost any part of programming language processing. What I tried to evoke with it is the most common notion of "compilation", i.e., statically compiling to a machine code. Maybe we should talk about _static compilation_ instead of just _compilation_? "Think of Nushell as a statically compiled language".